### PR TITLE
Fix backward compatibility issues with batch index acknowledgment.

### LINKED ISSUE #7460: Using 2.6.0 client with 2.5 cluster or older will lead to data loss for batched messages on redelivery
**Describe the bug**
In 2.6 client version the batch index acknowledge was added. But 2.5 cluster doesn't understand AckSet that was added for that and regards partially acknowleged batch as fully acknowledged. Therefore negatively acked messages will never be redelivered.

**To Reproduce**
Here is a sample program

```
package jtest;

import java.io.IOException;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.BitSet;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.logging.LogManager;

import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.Producer;
import org.apache.pulsar.client.api.Message;
import org.apache.pulsar.client.api.MessageId;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.PulsarClientException;
import org.apache.pulsar.client.api.Reader;
import org.apache.pulsar.client.api.SubscriptionInitialPosition;
import org.apache.pulsar.client.api.SubscriptionType;
import org.apache.pulsar.client.api.TypedMessageBuilder;
import org.apache.pulsar.client.impl.TopicMessageIdImpl;

/**
 * Hello world!
 */
public final class App {
    private static Instant start;

    private App() {
    }

    /**
     * Says hello to the world.
     *
     * @param args The arguments of the program.
     * @throws IOException
     */
    public static void main(String[] args) throws IOException {

        try {
            PulsarClient client = PulsarClient.builder().serviceUrl("pulsar://my-pulsar-cluster:30002").build();
            String topicName = "persistent://public/default/" + Instant.now().toEpochMilli();
            Consumer consumer = client.newConsumer()
                .topic(topicName)
                .subscriptionName("test-subscription")
                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                .negativeAckRedeliveryDelay(200, TimeUnit.MILLISECONDS)
                .subscribe();
            Producer<byte[]> producer = client.newProducer()
                    .topic(topicName)
                    .enableBatching(true)
                    .batchingMaxPublishDelay(100, TimeUnit.MILLISECONDS)
                    .create();

            producer.sendAsync(("Hello from Java 1").getBytes());
            producer.sendAsync(("Hello from Java 2").getBytes());

            Message msg1 = consumer.receive();
            System.out.printf("MessageId received: %s\n", msg1.getMessageId().toString());
            Message msg2 = consumer.receive();
            System.out.printf("MessageId received: %s\n", msg2.getMessageId().toString());
            consumer.acknowledge(msg1);
            consumer.negativeAcknowledge(msg2);

            Message msg3 = consumer.receive();
            System.out.printf("MessageId received: %s\n", msg3.getMessageId().toString());
            Message msg4 = consumer.receive();
            System.out.printf("MessageId received: %s\n", msg4.getMessageId().toString());
		} catch (PulsarClientException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }
}
```
in 2.5 client msg3 and msg4 will be redelivered and consumed, in 2.6 client they won't be consumed again.

**Expected behavior**
in 2.6 client with 2.5 cluster the behavior should be the same as 2.5 client.

**Additional context**
This makes it a breaking change, so clients shouldn't upgrade to 2.6 until cluster is also upgraded or this issue is fixed

cc @codelipenghui

### PR DESCRIPTION
**Pull request by codelipenghui**
Fixes #7460

### Motivation

Fix backward compatibility issues with batch index acknowledgment.

### Modifications

Disable batch index acknowledgment by default at the consumer side.

### Verifying this change

Integration tests added.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes)
  - The schema: (no)
  - The default values of configurations: ( no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by wolfstudy** [comment]
```
org.apache.pulsar.tests.integration.backwardscompatibility.SmokeTest2_4].testPublishAndConsume([pulsar://localhost:32774, false])-------[ERROR] Tests run: 24, Failures: 9, Errors: 0, Skipped: 3, Time elapsed: 234.916 s <<< FAILURE! - in TestSuite
[ERROR] pulsar-standalone-suite(org.apache.pulsar.tests.integration.backwardscompatibility.SmokeTest2_2)  Time elapsed: 3.4 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at org.junit.Assert.assertNotNull(Assert.java:722)
	at org.apache.pulsar.tests.integration.topologies.PulsarTestBase.testBatchIndexAckDisabled(PulsarTestBase.java:174)
	at org.apache.pulsar.tests.integration.backwardscompatibility.SmokeTest2_2.testBatchIndexAckDisabled(SmokeTest2_2.java:38)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

@codelipenghui please check the test case

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java [hunk 1]
```diff
@@ -609,6 +609,12 @@
      */
     ConsumerBuilder<T> enableRetry(boolean retryEnable);
 
+    /**
+     * Enable or disable the batch index acknowledgment. To enable this feature must ensure batch index acknowledgment
+     * feature is enabled at the broker side.
+     */
+    ConsumerBuilder<T> batchIndexAcknowledgmentEnabled(boolean batchIndexAcknowledgmentEnabled);
```
Comment (Lanayx) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java:616]: Shouldn't it be called **enableBatchIndexAcknowledgement** to conform other settings?

Comment (codelipenghui) [reply pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ConsumerBuilder.java:616]: Ok

**Pull request merged**